_posts/2019-07-12-mtls-with-nginx.md

```diff
@@ -0,0 +1,166 @@
+$ cd nginx-with-mtls-and-appserver
+$ docker-compose build
+```
+
+Thanks to Docker and Docker Compose, this is all we need to kick off the demo. The only final thing we need to do is to add the device that we will be connecting *from* to the whitelist. You can find out the WoTT Device ID by running `sudo wott-agent whoami` on the device you're connecting from.
+
+With the Device ID at hand, simply run the following commands on the server:
+
+```
+$ echo "MyDeviceId.d.wott.local" >> appserver/whitelist.txt
+```
+
+We can now fire up the server by simply running:
+
+```
+$ docker-compose up
+```
+
+To test the connection from your client, we need to find out two things:
+
+ * The IP of the server
+ * The WoTT Device ID of the server
+
+
+## Let's test the client
+
+Armed with the above information, we can now turn to our trusty old `curl` to test the setup.
+
+First, let's try connecting without passing on our certificate:
+
+```
+$ sudo curl \
+  --cacert /opt/wott/certs/ca.crt \
+  --resolve 'MyServerID.d.wott.local:443:192.168.X.Y' \
+  https://MyServerId.d.wott.local
+
+  <html>
+  <head><title>400 No required SSL certificate was sent</title></head>
+  <body>
+  <center><h1>400 Bad Request</h1></center>
+  <center>No required SSL certificate was sent</center>
+  <hr><center>nginx/1.16.0</center>
+  </body>
+  </html>
+
+```
+
+Because we have configured Nginx to require an SSL certificate, it will reject the connection, and you won't even be able to reach the appserver that we reverse proxy to.
+
+If we however pass on our certificate (and key), we are able to successfully access the appserver:
+
+```
+$ sudo curl \
+  --key /opt/wott/certs/client.key \
+  --cert /opt/wott/certs/client.crt \
+  --cacert /opt/wott/certs/ca.crt \
+  --resolve 'MyServerID.d.wott.local:443:192.168.X.Y' \
+  https://MyServerId.d.wott.local
+
+Access Granted!
+```
+
+## That's really cool, but what happened here?
+
+There's a lot to unpack in what we did above, so let's start with the client and work our way forward.
+
+
+### Client
+
+Let's break down the `curl` command we used to successfully connect to the server.
+
+These two lines tell `curl` to send the client certificate and key.
+```
+  --key /opt/wott/certs/client.key
+  --cert /opt/wott/certs/client.crt
+```
+This is our cryptographic identity that we will later use on the server for identification. It should however be said that the key is not sent to the server, but rather it is used for a cryptographic challenge (vastly simplified).
+
+Next, we need to tell `curl` to use the WoTT CA certificate to verify the remote server against (since WoTT is not a public CA):
+
+```
+  --cacert /opt/wott/certs/ca.crt
+```
+
+Lastly, we use a neat little feature in `curl` to tell it to map 'MyServerID.d.wott.local' to an IP address. We could instead have added this to our `/etc/hosts` file, but this is a quicker workaround when testing:
+
+```
+  --resolve 'MyServerID.d.wott.local:443:192.168.X.Y'
```

Comment:
just a note I've found not all IP addresses are in the 192.168.x.y form

Reply:
Yep this is because of [private networks](https://en.wikipedia.org/wiki/Private_network) as per RFC1918. I've updated it to make it more universal.